core/src/raw/oio/write/multipart_upload_write.rs

```diff
@@ -103,44 +107,116 @@ pub struct MultipartUploadPart {
     pub etag: String,
 }
 
+struct UploadFuture(BoxedFuture<Result<MultipartUploadPart>>);
+
+/// # Safety
+///
+/// wasm32 is a special target that we only have one event-loop for this UploadFuture.
+unsafe impl Send for UploadFuture {}
+
+/// # Safety
+///
+/// We will only take `&mut Self` reference for UploadFuture.
+unsafe impl Sync for UploadFuture {}
+
+impl Future for UploadFuture {
+    type Output = Result<MultipartUploadPart>;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.get_mut().0.poll_unpin(cx)
+    }
+}
+
+#[derive(Clone)]
+struct WriteTask {
+    part_number: usize,
+    bs: oio::ChunkedBytes,
+}
+
+/// # Safety
+///
+/// wasm32 is a special target that we only have one event-loop for this WriteTask.
+unsafe impl Send for WriteTask {}
+/// # Safety
+///
+/// We will only take `&mut Self` reference for WriteTask.
+unsafe impl Sync for WriteTask {}
+
 /// MultipartUploadWriter will implements [`Write`] based on multipart
 /// uploads.
 pub struct MultipartUploadWriter<W: MultipartUploadWrite> {
-    state: State<W>,
+    state: State,
+    w: Arc<W>,
 
-    cache: Option<oio::ChunkedBytes>,
     upload_id: Option<Arc<String>>,
     parts: Vec<MultipartUploadPart>,
+    pending: Option<WriteTask>,
+    futures: ConcurrentFutures<UploadFuture>,
+    next_part_number: usize,
 }
 
-enum State<W> {
-    Idle(Option<W>),
-    Init(BoxedFuture<(W, Result<String>)>),
-    Write(BoxedFuture<(W, Result<MultipartUploadPart>)>),
-    Close(BoxedFuture<(W, Result<()>)>),
-    Abort(BoxedFuture<(W, Result<()>)>),
+enum State {
+    Idle,
+    Init(BoxedFuture<Result<String>>),
+    Busy,
+    Close(BoxedFuture<Result<()>>),
+    Abort(BoxedFuture<Result<()>>),
 }
 
 /// # Safety
 ///
 /// wasm32 is a special target that we only have one event-loop for this state.
-unsafe impl<S: MultipartUploadWrite> Send for State<S> {}
+unsafe impl Send for State {}
 /// # Safety
 ///
 /// We will only take `&mut Self` reference for State.
-unsafe impl<S: MultipartUploadWrite> Sync for State<S> {}
+unsafe impl Sync for State {}
 
 impl<W: MultipartUploadWrite> MultipartUploadWriter<W> {
     /// Create a new MultipartUploadWriter.
-    pub fn new(inner: W) -> Self {
+    pub fn new(inner: W, concurrent: usize) -> Self {
         Self {
-            state: State::Idle(Some(inner)),
+            state: State::Idle,
 
-            cache: None,
+            w: Arc::new(inner),
             upload_id: None,
             parts: Vec::new(),
+            pending: None,
+            futures: ConcurrentFutures::new(1.max(concurrent)),
+            next_part_number: 0,
         }
     }
+
+    /// Increases part number and return the previous part number.
+    fn inc_part_number(&mut self) -> usize {
+        let part_number = self.next_part_number;
+        self.next_part_number += 1;
+        part_number
+    }
+
+    fn add_write_task(&mut self, bs: &dyn oio::WriteBuf) -> usize {
```

Comment:
I feel like we can still use the `cache` concept. And the function could be called as `fill_cache` or something.